lang/cs/Org.Apache.REEF.Network/Elastic/Topology/Logical/Impl/FlatTopology.cs

```diff
@@ -0,0 +1,448 @@
+        public IConfiguration GetTaskConfiguration(int taskId)
+        {
+            if (!_finalized)
+            {
+                throw new IllegalStateException("Cannot get task configuration from a not finalized topology.");
+            }
+
+            var confBuilder = TangFactory.GetTang().NewConfigurationBuilder();
+
+            if (taskId == _rootId)
+            {
+                var root = _nodes[_rootId];
+
+                foreach (var tId in root.Children)
+                {
+                    confBuilder.BindSetEntry<Config.OperatorParameters.TopologyChildTaskIds, int>(
+                        GenericType<Config.OperatorParameters.TopologyChildTaskIds>.Class,
+                        tId.TaskId.ToString(CultureInfo.InvariantCulture));
+                }
+            }
+
+            return confBuilder
+                .BindNamedParameter<Config.OperatorParameters.TopologyRootTaskId, int>(
+                    GenericType<Config.OperatorParameters.TopologyRootTaskId>.Class,
+                    _rootId.ToString(CultureInfo.InvariantCulture))
+                 .Build();
+        }
+
+        /// <summary>
+        /// This method is triggered when a node contacts the driver to synchronize the remote topology
+        /// with the driver's one.
+        /// </summary>
+        /// <param name="taskId">The identifier of the task asking for the update</param>
+        /// <param name="returnMessages">A list of message containing the topology update</param>
+        /// <param name="failureStateMachine">An optional failure machine to log updates</param>
+        public void TopologyUpdateResponse(
+            string taskId,
+            ref List<IElasticDriverMessage> returnMessages,
+            Optional<IFailureStateMachine> failureStateMachine)
+        {
+            if (taskId != _rootTaskId)
+            {
+                throw new IllegalStateException("Only root tasks are supposed to request topology updates.");
+            }
+
+            if (!failureStateMachine.IsPresent())
+            {
+                throw new IllegalStateException("Cannot update topology without failure machine.");
+            }
+
+            lock (_lock)
+            {
+                var list = _nodesWaitingToJoinTopology.ToList();
+                var update = new TopologyUpdate(_rootTaskId, list);
+                var data = new UpdateMessagePayload(
+                    new List<TopologyUpdate>() { update }, StageName, OperatorId, _iteration);
+                var returnMessage = new ElasticDriverMessageImpl(_rootTaskId, data);
+
+                returnMessages.Add(returnMessage);
+
+                if (_nodesWaitingToJoinTopology.Count > 0)
+                {
+                    Log.Log(Level.Info,
+                        "Tasks [{0}] are added to topology in iteration {1}",
+                        string.Join(",", _nodesWaitingToJoinTopology),
+                        _iteration);
```

Comment:
wrap this call into
```C#
if (Log.IsLoggable(Level.Info))
```
(call to `string.Join()` can be expensive)

Reply:
Yea. Fixed.